File: classpath/ikvm/internal/AssemblyClassLoader.java
/*
  Copyright (C) 2006-2013 Jeroen Frijters

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

  Jeroen Frijters
  jeroen@frijters.net
  
*/
// HACK because of historical reasons this class' source lives in ikvm/internal instead of ikvm/runtime
package ikvm.runtime;

import cli.System.Reflection.Assembly;
import gnu.java.util.EmptyEnumeration;
import ikvm.lang.Internal;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Enumeration;
import java.util.Vector;
import java.util.jar.Attributes;
import java.util.jar.Manifest;

public final class AssemblyClassLoader extends ClassLoader
{
    private final Assembly assembly;
    private boolean packagesDefined;

    public AssemblyClassLoader(Assembly assembly)
    {
        this(assembly, System.getSecurityManager());
    }

    // this constructor is used by the runtime to avoid the security check (by passing in null as the security manager)    
    AssemblyClassLoader(Assembly assembly, SecurityManager security)
    {
        super(null, security);
        this.assembly = assembly;
    }

    @Override
    protected Class loadClass(String name, boolean resolve) throws ClassNotFoundException
    {
        return LoadClass(this, assembly, name);
    }

    private static native Class LoadClass(ClassLoader classLoader, Assembly assembly, String name) throws ClassNotFoundException;

    @Override
    public URL getResource(String name)
    {
        return getResource(assembly, name);
    }

    @Override
    public Enumeration getResources(String name) throws IOException
    {
        return getResources(assembly, name);
    }

    @Override
    protected URL findResource(String name)
    {
        return getResource(assembly, name);
    }

    @Override
    protected Enumeration findResources(String name) throws IOException
    {
        return getResources(assembly, name);
    }

    private static native URL getResource(Assembly assembly, String name);
    
    private static native Enumeration getResources(Assembly assembly, String name) throws IOException;

    // also used by java.lang.LangHelper
    @Internal
    public static native String[] GetPackages(Assembly assembly);

    private static native URL GetManifest(Assembly assembly);

    private synchronized void lazyDefinePackagesCheck()
    {
        if(!packagesDefined)
        {
            packagesDefined = true;
            lazyDefinePackages();
        }
    }

    private static String getAttributeValue(Attributes.Name name, Attributes first, Attributes second)
    {
        String result = null;
        if(first != null)
        {
            result = first.getValue(name);
        }
        if(second != null && result == null)
        {
            result = second.getValue(name);
        }
        return result;
    }

    private Manifest getManifest()
    {
        try
        {
            URL url = GetManifest(assembly);
            if (url != null)
            {
                return new Manifest(url.openStream());
            }
        }
        catch (MalformedURLException _)
        {
        }
        catch (IOException _)
        {
        }
        return null;
    }

    private void lazyDefinePackages()
    {
        URL sealBase = getCodeBase();
        Manifest manifest = getManifest();
        Attributes attr = null;
        if(manifest != null)
        {
            attr = manifest.getMainAttributes();
        }
        String[] packages = GetPackages(assembly);
        for(int i = 0; i < packages.length; i++)
        {
            String name = packages[i];
            if(super.getPackage(name) == null)
            {
                Attributes entryAttr = null;
                if(manifest != null)
                {
                    entryAttr = manifest.getAttributes(name.replace('.', '/') + '/');
                }
                definePackage(name,
                    getAttributeValue(Attributes.Name.SPECIFICATION_TITLE, entryAttr, attr),
                    getAttributeValue(Attributes.Name.SPECIFICATION_VERSION, entryAttr, attr),
                    getAttributeValue(Attributes.Name.SPECIFICATION_VENDOR, entryAttr, attr),
                    getAttributeValue(Attributes.Name.IMPLEMENTATION_TITLE, entryAttr, attr),
                    getAttributeValue(Attributes.Name.IMPLEMENTATION_VERSION, entryAttr, attr),
                    getAttributeValue(Attributes.Name.IMPLEMENTATION_VENDOR, entryAttr, attr),
                    "true".equalsIgnoreCase(getAttributeValue(Attributes.Name.SEALED, entryAttr, attr)) ? sealBase : null);
            }
        }
    }

    @Override
    protected Package getPackage(String name)
    {
        lazyDefinePackagesCheck();
        return super.getPackage(name);
    }

    @Override
    protected Package[] getPackages()
    {
        lazyDefinePackagesCheck();
        return super.getPackages();
    }

    @Override
    public String toString()
    {
        return assembly.get_FullName();
    }

    private URL getCodeBase()
    {
        try
        {
            if(false) throw new cli.System.NotSupportedException();
            return new URL(assembly.get_CodeBase());
        }
        catch(cli.System.NotSupportedException _)
        {
        }
        catch(MalformedURLException _)
        {
        }
        return null;
    }
    
    // return the ClassLoader for the assembly. Note that this doesn't have to be an AssemblyClassLoader.
    public static native ClassLoader getAssemblyClassLoader(Assembly asm);
}

final class GenericClassLoader extends ClassLoader
{
    // this constructor avoids the security check in ClassLoader by passing in null as the security manager
    // to the IKVM specific constructor in ClassLoader
    GenericClassLoader()
    {
        super(null, null);
    }

    @Override
    public native String toString();

    @Override
    public URL getResource(String name)
    {
        Enumeration<URL> e = getResources(name);
        return e.hasMoreElements()
            ? e.nextElement()
            : null;
    }

    @Override
    public native Enumeration<URL> getResources(String name);

    @Override
    protected native URL findResource(String name);

    @Override
    protected Enumeration<URL> findResources(String name)
    {
        Vector<URL> v = new Vector<URL>();
        URL url = findResource(name);
        if (url != null)
        {
            v.add(url);
        }
        return v.elements();
    }
}

File: src/IKVM.Runtime/Java/Externs/ikvm/runtime/AssemblyClassLoader.cs
/*
  Copyright (C) 2002-2015 Jeroen Frijters

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

  Jeroen Frijters
  jeroen@frijters.net
  
*/
using System;
using System.Reflection;

using IKVM.Runtime;
using IKVM.Runtime;
using IKVM.Runtime.Accessors.Java.Lang;

using AssemblyClassLoader_ = IKVM.Runtime.AssemblyClassLoader;

namespace IKVM.Java.Externs.ikvm.runtime
{

    static class AssemblyClassLoader
    {

#if FIRST_PASS == false

        static ClassLoaderAccessor classLoaderAccessor;

        static ClassLoaderAccessor ClassLoaderAccessor => JVM.BaseAccessors.Get(ref classLoaderAccessor);

#endif

        public static void setWrapper(global::java.lang.ClassLoader _this, Assembly assembly)
        {
            ClassLoaderWrapper.SetWrapperForClassLoader(_this, AssemblyClassLoader_.FromAssembly(assembly));
        }

        public static global::java.lang.Class loadClass(global::java.lang.ClassLoader _this, string name, bool resolve)
        {
#if FIRST_PASS
            return null;
#else
            try
            {
                if (ClassLoaderAccessor.InvokeCheckName(_this, name) == false)
                    throw new ClassNotFoundException(name);

                var wrapper = (AssemblyClassLoader_)ClassLoaderWrapper.GetClassLoaderWrapper(_this);
                var tw = wrapper.LoadClass(name);
                if (tw == null)
                {
                    Tracer.Info(Tracer.ClassLoading, "Failed to load class \"{0}\" from {1}", name, _this);
                    global::java.lang.Throwable.suppressFillInStackTrace = true;
                    throw new global::java.lang.ClassNotFoundException(name);
                }
                Tracer.Info(Tracer.ClassLoading, "Loaded class \"{0}\" from {1}", name, _this);
                return tw.ClassObject;
            }
            catch (ClassNotFoundException x)
            {
                Tracer.Info(Tracer.ClassLoading, "Failed to load class \"{0}\" from {1}", name, _this);
                throw new global::java.lang.ClassNotFoundException(x.Message);
            }
            catch (ClassLoadingException x)
            {
                Tracer.Info(Tracer.ClassLoading, "Failed to load class \"{0}\" from {1}", name, _this);
                throw x.InnerException;
            }
            catch (RetargetableJavaException x)
            {
                Tracer.Info(Tracer.ClassLoading, "Failed to load class \"{0}\" from {1}", name, _this);
                throw x.ToJava();
            }
#endif
        }

        public static global::java.net.URL getResource(global::java.lang.ClassLoader _this, string name)
        {
#if !FIRST_PASS
            var wrapper = (AssemblyClassLoader_)ClassLoaderWrapper.GetClassLoaderWrapper(_this);
            foreach (global::java.net.URL url in wrapper.GetResources(name))
                return url;
#endif
            return null;
        }

        public static global::java.util.Enumeration getResources(global::java.lang.ClassLoader _this, string name)
        {
#if FIRST_PASS
            return null;
#else
            return new global::ikvm.runtime.EnumerationWrapper(((AssemblyClassLoader_)ClassLoaderWrapper.GetClassLoaderWrapper(_this)).GetResources(name));
#endif
        }

        public static global::java.net.URL findResource(global::java.lang.ClassLoader _this, string name)
        {
#if !FIRST_PASS
            AssemblyClassLoader_ wrapper = (AssemblyClassLoader_)ClassLoaderWrapper.GetClassLoaderWrapper(_this);
            foreach (global::java.net.URL url in wrapper.FindResources(name))
            {
                return url;
            }
#endif
            return null;
        }

        public static global::java.util.Enumeration findResources(global::java.lang.ClassLoader _this, string name)
        {
#if FIRST_PASS
            return null;
#else
            return new global::ikvm.runtime.EnumerationWrapper(((AssemblyClassLoader_)ClassLoaderWrapper.GetClassLoaderWrapper(_this)).FindResources(name));
#endif
        }

#if !FIRST_PASS
        static global::java.net.URL GetCodeBase(Assembly assembly)
        {
            try
            {
                return new global::java.net.URL(assembly.CodeBase);
            }
            catch (NotSupportedException)
            {

            }
            catch (global::java.net.MalformedURLException)
            {

            }

            return null;
        }

        /// <summary>
        /// Returns the first non-null value for the attribute in either the first attribute set or the second.
        /// </summary>
        /// <param name="name"></param>
        /// <param name="first"></param>
        /// <param name="second"></param>
        /// <returns></returns>
        static string GetAttributeValue(global::java.util.jar.Attributes.Name name, global::java.util.jar.Attributes first, global::java.util.jar.Attributes second)
        {
            return first?.getValue(name) ?? second?.getValue(name);
        }

#endif

#if !FIRST_PASS

        /// <summary>
        /// Gets the JAR manifest for the given assembly and resource name.
        /// </summary>
        /// <param name="assembly"></param>
        /// <param name="resourceName"></param>
        /// <returns></returns>
        /// <exception cref="ArgumentNullException"></exception>
        /// <exception cref="ArgumentException"></exception>
        static global::java.util.jar.Manifest GetManifestForAssemblyJar(Assembly assembly, string resourceName)
        {
            if (assembly is null)
                throw new ArgumentNullException(nameof(assembly));
            if (string.IsNullOrEmpty(resourceName))
                throw new ArgumentException($"'{nameof(resourceName)}' cannot be null or empty.", nameof(resourceName));

            using var resStream = assembly.GetManifestResourceStream(resourceName);
            if (resStream == null)
                throw new InternalException($"Could not find assembly resource {resourceName}.");

            using var wrpStream = new global::ikvm.io.InputStreamWrapper(resStream);
            using var jarStream = new global::java.util.jar.JarInputStream(wrpStream);
            var manifest = jarStream.getManifest();

            return manifest;
        }

#endif

        /// <summary>
        /// Defines all of the packages in the class loader that are not currently defined, using the assembly as a source.
        /// </summary>
        /// <param name="_this"></param>
        public static void lazyDefinePackages(global::java.lang.ClassLoader _this)
        {
#if !FIRST_PASS
            var wrapper = (AssemblyClassLoader_)ClassLoaderWrapper.GetClassLoaderWrapper(_this);
            var sealBase = GetCodeBase(wrapper.MainAssembly);

            foreach (var packages in wrapper.GetPackageInfo())
            {
                // if the package exists in a JAR within the main assembly, load its manifest
                var manifest = packages.Key != null ? GetManifestForAssemblyJar(wrapper.MainAssembly, packages.Key) : null;

                // if we loaded a manifest, get the main attributes
                var mainAttributes = manifest?.getMainAttributes();

                // for each package, check if defined, else define
                foreach (var name in packages.Value)
                {
                    if (_this.getPackage(name) != null)
                        continue;

                    // manifest might contain an overriding entry for this specific package
                    var packageAttributes = manifest?.getAttributes(name.Replace('.', '/') + '/');

                    // define package with package information from package attributes or main attributes
                    _this.definePackage(
                        name,
                        GetAttributeValue(global::java.util.jar.Attributes.Name.SPECIFICATION_TITLE, packageAttributes, mainAttributes),
                        GetAttributeValue(global::java.util.jar.Attributes.Name.SPECIFICATION_VERSION, packageAttributes, mainAttributes),
                        GetAttributeValue(global::java.util.jar.Attributes.Name.SPECIFICATION_VENDOR, packageAttributes, mainAttributes),
                        GetAttributeValue(global::java.util.jar.Attributes.Name.IMPLEMENTATION_TITLE, packageAttributes, mainAttributes),
                        GetAttributeValue(global::java.util.jar.Attributes.Name.IMPLEMENTATION_VERSION, packageAttributes, mainAttributes),
                        GetAttributeValue(global::java.util.jar.Attributes.Name.IMPLEMENTATION_VENDOR, packageAttributes, mainAttributes),
                        string.Equals(GetAttributeValue(global::java.util.jar.Attributes.Name.SEALED, packageAttributes, mainAttributes), "true", StringComparison.OrdinalIgnoreCase) ? sealBase : null);
                }
            }
#endif
        }

        public static string toString(global::java.lang.ClassLoader _this)
        {
            return ((AssemblyClassLoader_)ClassLoaderWrapper.GetClassLoaderWrapper(_this)).MainAssembly.FullName;
        }

        public static global::java.lang.ClassLoader getAssemblyClassLoader(Assembly asm)
        {
            // note that we don't do a security check here, because if you have the Assembly object,
            // you can already get at all the types in it.
            return AssemblyClassLoader_.FromAssembly(asm).GetJavaClassLoader();
        }

    }

}
